src/native/fs/fs_napi.cpp

```diff
@@ -158,6 +202,110 @@ set_fs_worker_stats(Napi::Env env, Napi::Object fs_worker_stats, std::string wor
 }
 
+static int 
+get_single_user_xattr(int fd, std::string key, std::string &value) {
+    #ifdef __APPLE__
+        ssize_t value_len = fgetxattr(fd, key.c_str(), NULL, 0, 0, 0);
+    #else
+        ssize_t value_len = fgetxattr(fd, key.c_str(), NULL, 0);
+    #endif
+
+    if (value_len == -1) {
+        return -1;
+    }
+
+    if (value_len > 0) {
+        Buf val(value_len);
+
+        #ifdef __APPLE__
+            value_len = fgetxattr(fd, key.c_str(), val.cdata(), value_len, 0, 0);
+        #else
+            value_len = fgetxattr(fd, key.c_str(), val.cdata(), value_len);
+        #endif
+
+        if (value_len == -1) {
+            return -1;
+        }
+        
+        value = std::string(val.cdata(), value_len);
+    } else if (value_len == 0) {
+        value = "";
+    }
+    return 0;
+}
+
+static int 
+get_fd_xattr(int fd, XattrMap &xattr, bool skip_user_xattr)
+{
+    if (skip_user_xattr) { // we won't list the attributes just return the prefefined list
+        for (std::string key: USER_XATTRS) {
+            std::string value;
+            int rc =  get_single_user_xattr(fd, key, value);
+            if (rc) return rc;            
+            xattr[key] = value;
+        }
+    } else {
+        #ifdef __APPLE__
+            ssize_t buf_len = flistxattr(fd, NULL, 0, 0);
+        #else
+            ssize_t buf_len = flistxattr(fd, NULL, 0);
+        #endif
+
+        if (buf_len <= 0) {
+            return buf_len;
+        }
+
+        Buf buf(buf_len);
+
+        #ifdef __APPLE__
+            buf_len = flistxattr(fd, buf.cdata(), buf_len, 0);
+        #else
+            buf_len = flistxattr(fd, buf.cdata(), buf_len);
+        #endif
+
+        // No xattr, nothing to do  
+        if (buf_len <= 0) {
+            return buf_len;
+        }
+
+        while (buf.length()) {
+            std::string key(buf.cdata());
+            std::string value;
+            int rc =  get_single_user_xattr(fd, key, value);
+            if (rc) return rc;            
+            xattr[key] = value;
+            buf.slice(key.size() + 1, buf.length());
+        }
+    }
+    return 0;
+}
+
+static int 
+get_fd_gpfs_xattr(int fd, XattrMap &xattr)
+{
+    for (std::string key: GPFS_XATTRS) {
+        gpfsRequest_t gpfsGetXattrRequest;
+        build_gpfs_get_ea_request(&gpfsGetXattrRequest, key);
+        int rc = dlsym_gpfs_fcntl(fd, &gpfsGetXattrRequest);
+        if (rc != 0)
+        { 
+            return rc;
+        }
+        int error_code = gpfsGetXattrRequest.payload.errReasonCode;
+        if (error_code == GPFS_FCNTL_ERR_NONE)
+        {
+            int name_len = gpfsGetXattrRequest.payload.nameLen;
+            int buffer_len = gpfsGetXattrRequest.payload.bufferLen;
+            xattr[key] = std::string(gpfsGetXattrRequest.buffer[ROUNDUP(name_len, 8)], buffer_len);
+        } else if (error_code == GPFS_FCNTL_ERR_NO_ATTR) {
+            xattr[key] = "";
+        } else {
+            return error_code;
```

Comment:
maybe add a specific log here so that we know this error code comes from gpfs and not from posix?